### PR TITLE
Add AdamW optimizer to extension/training (carries #18848 by @BryanBradfo)

### LINKED ISSUE #18766: Support for AdamW optimizer in ExecuTorch
As of now, ExecuTorch supports SGD as the only optimizer, with Adam listed as planned.

Given that AdamW is the default optimizer used in most modern PyTorch training workflows, is AdamW support also expected as part of the ExecuTorch optimizer roadmap?

Are there any known design or semantic considerations in existing AdamW implementations (e.g., weight decay behavior or optimizer state handling) that influence how AdamW support is approached in ExecuTorch?

Thanks!

cc @JacobSzwejbka

### PR DESCRIPTION
**Pull request by psiddh**
> ## Attribution
>
> **This is Bryan's work, not mine.** The AdamW implementation, the test suite, and the
> design decisions in this PR are all from **@BryanBradfo** in **#18848**, opened 13 April
> 2026. I am carrying it forward only because that PR cannot merge: two of its commits
> (`5f1c90c`, `bda405f`) are not covered by a signed Linux Foundation CLA, and the
> contributor has not responded to requests to sign it (pinged 2 Sept, 8 Sept, and 15 Sept).
>
> EasyCLA keys off commit authorship, so re-opening with their commits intact fails the same
> check. Squashing under my authorship is what makes the check pass — maintainers should be
> aware that this is the *only* reason the commit says what it says, and should decide
> whether they are comfortable landing it on that basis rather than waiting for
> @BryanBradfo to sign. I have no objection to this being closed in favour of #18848 if the
> CLA gets signed.
>
> Original PR: https://github.com/pytorch/executorch/pull/18848

Fixes #18766

## Summary

Adds AdamW to the training optimizer extension. It is a port of the existing SGD
implementation at `extension/training/optimizer/sgd.{h,cpp}`, with the main algorithmic
difference being decoupled weight decay — the parameter is decayed directly instead of
mixing the decay into the gradient. Matches `torch.optim.AdamW` with default settings.

## Scope

C++ only. Python bindings are left out on purpose: the pybindings file has a TODO to build
a generic optimizer interface first, so copying `PySGD` to `PyAdamW` now would just add
duplication. `amsgrad` and `maximize` are also left out — both rarely used and easy to add
later.

## Changes on top of #18848

Two things, both mine:

- **Build fix.** `make_tensor_ptr` gained a `Device` parameter after #18848 was written, so
  the two state-buffer calls were passing `TensorShapeDynamism::STATIC` where a `Device` is
  expected and the deleter where the dynamism is expected. This is what broke every build
  job on that PR. Both moment buffers come from `malloc`, so they get an explicit CPU device.
- **`malloc` null check** before `memset`, as flagged in review on #18848. The two
  allocations were used unconditionally, so a failure on a memory-constrained target
  dereferenced null instead of returning `Error::MemoryAllocationFailed`.

The other review points on #18848 (dtype/shape validation, hyperparameter range checks) are
deliberately not addressed here. @BryanBradfo's argument was that SGD has neither and adding
them to AdamW alone would be inconsistent, and @JacobSzwejbka noted the Copilot review is
noisy. Both optimizers can be hardened together in a follow-up.

## Verification

`lintrunner` is clean, and `adamw.cpp` compiles against current `main` — I reproduced the
exact CI error from #18848 locally and confirmed this clears it.

I could not execute `adamw_test` (this machine SIGKILLs freshly built binaries), so instead
I transcribed `AdamW::step`'s arithmetic verbatim, including the float round-trips, and
diffed it against `torch.optim.AdamW` over six configurations:

```
case                                adamw.cpp        torch     absdiff
Simple  (wd=0, g=-1, 10 steps)      2.0000002    2.0000002    0.00e+00
DecoupledWD (g=0, wd=0.5, 1)        0.9500000    0.9500000    0.00e+00
MultiParam p1 (g=-1, 5 steps)       1.5000001    1.5000001    0.00e+00
MultiParam p2 (g=+1, 5 steps)       1.4999999    1.4999999    0.00e+00
Both wd and grad nonzero            2.9613018    2.9613023    4.77e-07
Nondefault betas                   -0.2605234   -0.2605233    2.98e-08
```

The decoupled-weight-decay ordering matches and all four gtest expectations agree with
PyTorch independently. @BryanBradfo reported `buck2 test //extension/training/optimizer/test:adamw_test`
passing 6/6 and the SGD regression staying green on #18848; someone with a working Buck
setup should confirm that still holds, since the tests are a Buck-only target and CI does
not run them.

cc @JacobSzwejbka

---

The build fix and this PR were prepared with Claude Code.

### CODE REVIEW

**Comment by psiddh** [comment]
Dispositioning the three Copilot findings. All three were also raised on #18848; the `make_tensor_ptr` and `malloc` ones it flagged there are gone because those are fixed.

**`Cpp` -> `C++` in the README** — fixed, it was right.

**Constant-gradient tests** — fair, and fixed. A constant ±1 gradient makes Adam's normalized update very nearly sign-only, so the existing three tests would have passed with a broken moment recurrence or bias correction. Added `AdamWOptimizerVaryingGradient`: an eight-step sequence `{0.5, -1.5, 2.0, -0.25, 1.0, 0.75, -2.5, 0.1}`, run with and without decoupled weight decay, expectations from `torch.optim.AdamW` at `1e-6`. Adam with coupled L2 lands on `0.79996` against AdamW's `0.77575` on the decay case, so the tolerance also pins the AdamW-vs-Adam distinction under a non-trivial gradient.

**dtype/layout/shape validation (flagged High)** — not fixed here, deliberately. The report is technically correct: `const_data_ptr<T>()` is an unchecked `static_cast` with no dtype enforcement, so a non-Float tensor would reinterpret bytes, and a `p.numel() != g.numel()` mismatch would read past the state buffers. But `sgd.cpp`'s `add_out_hack` / `mul_out_hack` / `clone_out_hack` have exactly the same hazard on the same public `step()` surface — this is a property of the optimizer extension as a whole, not something AdamW introduces. @BryanBradfo made this argument on #18848 and I agree with it. Hardening one optimizer and not the other would leave the codebase in a worse-documented state than either fixing both or fixing neither. Happy to do both in a follow-up if @JacobSzwejbka wants it.

Content is byte-identical to #18848 (tree `b580cf4`), which has the same two commits pushed to it.

**Comment by psiddh** [comment]
Copilot's follow-up on the allocation-failure path was right, and it was on code I added rather than @BryanBradfo's. My `malloc` null check returned `MemoryAllocationFailed` after the in-place weight decay had already changed `p`, while its moments and step counter stayed put — so a caller retrying the failed step would decay the parameter a second time. Before my change there was no early return there at all, so I introduced that window.

Fixed by moving the state lookup and allocation above the decay, so nothing mutates `p` until the fallible part has succeeded. Decay still precedes the moment update, so the arithmetic is untouched and the reference values in `AdamWOptimizerVaryingGradient` are unchanged.

Same commit is on #18848; trees still identical at `9b843c2`.

**Review by JacobSzwejbka — APPROVED** [review]

**Comment by psiddh** [comment]
## Upgraded verification: the real C++ now runs, not just a transcription

Earlier I could only say I had transcribed `AdamW::step()`'s arithmetic and diffed it against
`torch.optim.AdamW`, because the gtest binary was being SIGKILLed in my environment. I got the
actual compiled ExecuTorch code running, so this is now direct evidence.

```
case                                    got     expected     |diff|  result
Simple (wd=0,g=-1,x10)           2.00000024   2.00000000   2.38e-07  PASS
DecoupledWD (g=0,wd=.5,x1)       0.94999999   0.94999999   0.00e+00  PASS
MultiParam p1 (g=-1,x5)          1.50000012   1.50000000   1.19e-07  PASS
MultiParam p2 (g=+1,x5)          1.49999988   1.50000000   1.19e-07  PASS
Varying wd=0 (8 steps)           0.84547687   0.84547687   0.00e+00  PASS
Varying wd=0.1 (8 steps)         0.77574724   0.77574724   0.00e+00  PASS
6/6 passed
```

The two varying-gradient values are the ones taken from `torch.optim.AdamW`, and the real
implementation reproduces them exactly.

Per-parameter state isolation, which the cases above do not cover because they use one optimizer
each, checked separately with two parameters in a single optimizer:

```
two params, one optimizer, 5 steps:
  a: 1.50000012 (expect ~1.5) PASS
  b: 1.49999988 (expect ~1.5) PASS
  partial gradients: b unchanged = PASS (1.49999988 -> 1.49999988)
```

That last line also covers the `named_gradients.find(...) == end()` skip path, and the whole run
exercises the state-allocation reorder from the third commit, since state is allocated on step 1
and reused on steps 2-5.

### What this still does not cover

The gtest harness itself (`TensorFactory`, the `ET_EXPERIMENTAL` wrappers) and the Buck target. My
driver calls `AdamW::step()` directly against `TensorImpl`. The gtest binary is still SIGKILLed
here by what looks like an endpoint-security rule — a trivial binary linking the same objects runs
fine, so it is not the code. Someone should still run
`buck2 test //extension/training/optimizer/test:adamw_test` before merge.

**Comment by psiddh** [comment]
Closing this PR in favor of #22916
